collection-scripts/gather_monitoring

```diff
@@ -1,29 +1,208 @@
+        get cm default-ingress-cert \
+        -o jsonpath='{.data.ca-bundle\.crt}' > "${CA_BUNDLE}"
+}
+
+function init {
+
+    mkdir -p "${PROM_PATH}"
+
+    get_ca_bundle
+
+    get_env
+
+    # Query Param: 'start' and 'end' timestamp
+    local _START_HUMAN=${GATHER_MONIT_START_DATE:-${GATHER_MONIT_START_DATE_DEFAULT}}
+    local _END_HUMAN=${GATHER_MONIT_END_DATE:-${GATHER_MONIT_END_DATE_DEFAULT}}
+    declare -g DATE_START=$(date -d "${_START_HUMAN}" +%s)
+    declare -g DATE_END=$(date -d "${_END_HUMAN}" +%s)
+
+    # Query Param: 'step'. Low metric resolution with long range could be limited to 11k datapoints.
+    declare -g QUERY_STEP=${GATHER_MONIT_QUERY_STEP:-${GATHER_MONIT_QUERY_STEP_DEFAULT}}
+
+    declare -g PROM_API_QSTR="start=${DATE_START}&end=${DATE_END}&step=${QUERY_STEP}"
+
+    cat <<-EOF
+INFO: Metrics config time range from=${DATE_START} to=${DATE_END} step=${QUERY_STEP}
+INFO: Metrics human  time range from=[$(date -d "@${DATE_START}")] to [$(date -d "@${DATE_END}")]
+INFO: Config Env GATHER_MONIT_START_DATE: ${GATHER_MONIT_START_DATE:-""}
+INFO: Config Env GATHER_MONIT_END_DATE: ${GATHER_MONIT_END_DATE:-""}
+INFO: Config Env GATHER_MONIT_QUERY_STEP: ${GATHER_MONIT_QUERY_STEP:-""}
+INFO: Config Env GATHER_MONIT_CUSTOM_METRICS: ${GATHER_MONIT_CUSTOM_METRICS:-""}
+INFO: Config Env GATHER_MONIT_DISCOVERY_PREFIXES: ${GATHER_MONIT_DISCOVERY_PREFIXES:-""}
+INFO: Prometheus endpoint: https://${PROM_HOST}
+INFO: Prometheus base query: ${PROM_API}/<uri>?query=<metric>&${PROM_API_QSTR}"
+EOF
+}
+
+function cleanup {
+    rm "${CA_BUNDLE}"
+}
+
+function oc_get {
+    oc get \
+        --certificate-authority="${CA_BUNDLE}" \
+        --token="${SS_TOKEN}" \
+        "$@"
+}
+
+function prom_get_api {
+    local API_URI="${PROM_API}/${1}"; shift
+
+    oc_get \
+        --server="https://${PROM_HOST}" \
+        --raw="${API_URI}"
+}
+
+function prom_get_query_range {
+    local METRIC="${1}"; shift
+    local FILE_OK="${PROM_PATH}/metric-${METRIC}.json.gz";
+    local FILE_ERR="${PROM_PATH}/metric-${METRIC}.stderr";
+    local API_URI="query_range?query=${METRIC}&${PROM_API_QSTR}"
+
+    echo "INFO: Collecting metric /query_range?query=${METRIC}"
+    prom_get_api "${API_URI}" 2> "${FILE_ERR}" | gzip > "${FILE_OK}"
+}
+
+#
+# Gather specific metrics by name.
+# Env var: GATHER_MONIT_CUSTOM_METRICS
+#
+function get_custom_metrics {
+
+    if [[ -z "${GATHER_MONIT_CUSTOM_METRICS:-}" ]]; then
+        return
+    fi
+    echo "INFO: Collecting custom metrics..."
+
+    for METRIC in ${GATHER_MONIT_CUSTOM_METRICS}; do
+        prom_get_query_range "${METRIC}"
+    done
+}
+
+#
+# Discovery and gather metrics by prefixes.
+# Env var: GATHER_MONIT_DISCOVERY_PREFIXES
+#
+function get_discovery_metrics {
+
+    if [[ -z "${GATHER_MONIT_DISCOVERY_PREFIXES:-}" ]]; then
+        return
+    fi
+    echo "INFO: Collecting metrics by prefixes..."
+
+    prom_get_api "label/__name__/values" \
+        | python2 -c \
```

Comment:
is it better to just gather the data and do the parsing and extraction later / offline?

Reply:
@sthaha it was used in alternative of `jq`, that I deprecated it in previous review. It used online on the script to discovery the metrics (next loop).